File: leetcode/101/src/P101.java
/**
 * Created by qwordy on 2/22/16.
 * 101. Symmetric Tree
 */

class TreeNode {
	int val;
	TreeNode left;
	TreeNode right;
	TreeNode(int x) { val = x; }
}

class Solution {
	public boolean isSymmetric(TreeNode root) {
		return root == null || sym(root.left, root.right);
	}

	private boolean sym(TreeNode p, TreeNode q) {
		return p == null && q == null || p != null && q != null && p.val == q.val && sym(p.left, q.right) && sym(p.right, q.left);
	}
}

public class P101 {
}

File: leetcode/101/P101.rb
# 101. Symmetric Tree
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Boolean}
def is_symmetric(root)
	!root || sym(root.left, root.right) ? true : false
end

def sym p, q
	!p && !q || p && q && p.val == q.val && sym(p.left, q.right) && sym(p.right, q.left)
end

